lib/IRGen/IRGenSIL.cpp

```diff
@@ -2053,38 +2053,87 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
+    std::string argumentName = std::get<0>(argumentNameAndLowering);
+    auto argumentLowering = std::get<1>(argumentNameAndLowering);
+    LLVM_DEBUG(llvm::dbgs() << "IRGen graph_op argument " << argumentName
+               << "with lowering " << (unsigned)argumentLowering);
+    switch (argumentLowering) {
+    case GraphOperationInfo::ArgumentLowering::Out:
+      assert(0 && "Attributes cannot be output parameters");
+    case GraphOperationInfo::ArgumentLowering::UnknownShapeListAttribute:
+      assert(0 && "UnknownShapeListAttribute should have been eliminated "
+                  "by deabstraction");
+    case GraphOperationInfo::ArgumentLowering::TypeListAttribute:
+      assert(0 && "TypeListAttribute should have been eliminated by "
+                  "deabstraction");
+    case GraphOperationInfo::ArgumentLowering::TensorAttribute:
+      // See GraphOperationInfo.h for more information on why TensorAttribute
+      // cannot be dynamic.
+      assert(0 && "TensorAttributes cannot be dynamic");
+    case GraphOperationInfo::ArgumentLowering::Input: {
+      assert(argumentName.empty() && "inputs cannot have names");
+      switch (structuredArgument.getKind()) {
+      case GraphOperationInfo::SAK_Single: {
+        LLVM_DEBUG(llvm::dbgs() << " Adding a single input.\n");
+        auto tensorHandle = structuredArgument.getSingleArgument();
+        addTensorhandleAsOpInput(tensorHandle);
+        break;
+      }
+      case GraphOperationInfo::SAK_List: {
+        LLVM_DEBUG(llvm::dbgs()
+                   << " Adding input list of size "
+                   << structuredArgument.getArgumentList().size() << ".\n");
+        for (auto tensorHandle : structuredArgument.getArgumentList()) {
+          addTensorhandleAsOpInput(tensorHandle);
+        }
+        // Set special attr to represent the list length, if needed. This is not
+        // needed in graph lowering.
+        if (auto numberAttr = inputListNumberAttr(opInfo.getOperationName())) {
+          auto setAttrFn = IGM.getTFE_OpSetAttrIntFn();
+          auto listLenAddr = createStringValAddr(IGM, numberAttr);
+          auto listLenVal = llvm::ConstantInt::get(
+              IGM.Int64Ty, structuredArgument.getArgumentList().size());
+          Builder.CreateCall(setAttrFn, {op, listLenAddr, listLenVal});
+        }
+      }
+      }
       break;
     }
-    case GraphOperationInfo::SAK_List: {
-      LLVM_DEBUG(llvm::dbgs()
-                 << " Adding input list of size "
-                 << structuredArgument.getArgumentList().size() << ".\n");
-      for (auto tensorHandle : structuredArgument.getArgumentList()) {
-        addTensorhandleAsOpInput(tensorHandle);
+    case GraphOperationInfo::ArgumentLowering::TFDataTypeAttribute: {
+      switch (structuredArgument.getKind()) {
+      case GraphOperationInfo::SAK_Single: {
+        auto silValue = structuredArgument.getSingleArgument();
+        StringRef typeString = silValue->getType().getASTType().getString();
```

Comment:
Isn't there a more direct way to check if this is a Builtin.Int32 type?

Reply:
done